### ISSUE TITLE
Automatic schema maker doesn't detect type for edge properties

### ISSUE THREAD

**Issue by francescoinfante** [JanusGraph/janusgraph#1066]
On JanusGraph 0.2.0:

If I create a new property key by adding it to a node, automatically a new property key is added to the schema with the correct datatype (according to https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphDefaultSchemaMaker.java)

If I create a new property key by adding it to an edge, a new property key is added to the schema with the class `Object.class`.

**Comment by palwe-prafulla** [comment]
I see similar issue in 0.2.0 on my set up.
i create edge with property value of type Integer then I call addEdge API to add edge on from vertex and then I ad proprties use edge.property()
The input properties map is clearly indicating property type as Integer.
However as soon as I do get property using edge.property after setting it, I get the type as String.
Not sure if any automatic type conversion is carried out with Java APIs for Janusgraph.

Strangely I dont see this issue for Long data type on the edge for some reason.
This is really wierd.

Any comments?

**Comment by porunov** [comment]
@palwe-prafulla Please, upgrade your JanusGraph instances to more recent versions of JanusGraph.
This issue was fixed in the version 0.2.1 of JanusGraph. You can easily upgrade to 0.2.3 version from 0.2.0 version.
Also, notice that 0.2 branch is EOL and isn't supported anymore, so any bugs found in 0.2.x versions will not be fixed anymore.
If you want to receive bugfixes and improvments I would suggest to upgrade your instances to the latest JanusGraph version (currently it is 0.4.0). If you cannot upgrade your instances to 0.4.0 because of usage of backend versions which are EOL, you may consider to upgrade to 0.3.x version, which is still supported by JanusGraph community. Notice that schema of `0.3.x` and `0.4.x` is different from `0.2.x`, please refer to upgrade instructions for more details: https://docs.janusgraph.org/changelog/#upgrade-instructions_1

**Comment by palwe-prafulla** [comment]
Thanks @porunov for your comment and info.
I just wanted to post and update about my analysis since last time I observed the wierd behavior.

It looks like if you reuse the same property key between multiple types of vertices or edges with different value types then janusgraph automatically assign the value type when the first time this property is used.
Thats the reason for my case the property was converted to String even if the value type was Integer because it was used by a vertex earlier as a String type value.
Thats also the reason I dont see issues with other edges with value type Int as those properties are not used by anything else.

I want to check if this behavior is fixed in 0.2.1 or other releases later or do we still have to use different property names if the value types are different?
Basically if we reuse the same property key between multiple types of vertices/edges, would this problem still occur in 0.2.1 or other releases?

**Comment by porunov** [comment]
> It looks like if you reuse the same property key between multiple types of vertices or edges with different value types then janusgraph automatically assign the value type when the first time this property is used.

Partially correct. If you didn't disable automatic schema maker, then the type will be assigned on the first use. https://docs.janusgraph.org/basics/schema/#automatic-schema-maker
Please see: You can disable this feature and explicitly create your schema.

> Thats the reason for my case the property was converted to String even if the value type was Integer because it was used by a vertex earlier as a String type value.

Of course, the property may have only a single type. If you already used a property somewhere with a different type, you may need to create another property with necessary type. Notice that each schema elements must have **unique** names. It might be not obvious if you used a database previously where properties with the same name may very in types from table to table (something like MySQL).
If you used a type `Long` to for a property `age`, you will not be able to use type `String` for the same property anymore even if you are using this property in another vertex / edge. I.e. if you need a property which stores age in different types for different vertices / edges, you may want to create a more specific property. For example: `user_age` - `Long`, `earth_age` - `String`, `friendship_age` - `Integer`, etc. 
Please see: https://docs.janusgraph.org/basics/schema/#defining-property-keys

> I want to check if this behavior is fixed in 0.2.1 or other releases later or do we still have to use different property names if the value types are different?

This issue isn't related to the the issue you are describing. You will still have to use different property names if the types are different. This is a requirement for schema creation.

> Basically if we reuse the same property key between multiple types of vertices/edges, would this problem still occur in 0.2.1 or other releases?

This isn't a problem. Please, see documentation links I've posted above. Yes, this behaviour will occur in the future releases but it is an expected and correct behaviour.

**Comment by palwe-prafulla** [comment]
@porunov 
Thanks a lot for more information on this and explaining the behavior.
I understand that I may have to use different property names for different types even though context or symantics are same.

I have tested with different property names and it works fine.
This original thread explains the issue of schema maker not recognizing the value types on edge properties and now I am confused because if I use different property on the edge then it still gives back the correct type. So how is the original issue which was posted about value types being given back as Object type instead of specific type is even an issue?

**Comment by porunov** [comment]
@palwe-prafulla I was not related neither to this issue nor to the fix of this issue, so I can't provide more information regarding this issue. Also, as I said, janusgraph versions `0.2.x` are not supported anymore, so I don't see any reason to go back and check some functionality in the `0.2.x` versions.
In addition, if you are satisfied with `0.2.0` version of JanusGraph, you don't need to upgrade you instances.

**Comment by palwe-prafulla** [comment]
Thanks @porunov
